### DISCUSSION TITLE
Missing symbols for most imports (not app, auth, database)

### DISCUSSION THREAD

**Discussion by akauppi** [firebase/firebase-js-sdk#4534] [Archive] [answered]
On WebStorm 2020.3, only three of the module namespaces have symbols available to the IDE:

<img width="422" alt="image" src="https://user-images.githubusercontent.com/1297734/109194360-0065a200-77a2-11eb-8107-149412f391f6.png">

I looked briefly into the packaging, but wasn't able to figure out the cause for this. The damage to developer experience is real, since we often use Cmd-B (`Declaration & Usage`) for seeing information about a class or function. For the ones with missing symbols, this is not available.

firebase 0.900.15

**Comment by Feiyang1** [comment]
That's weird. Is it possible you still have prod version of the SDK installed? Can you please try to delete `node_modules` and reinstall to see if the issue goes away? If it persists, can you share your `package.json`?

**Reply by akauppi** [reply]
I did that. And have cleared IDE caches.

I'll attach the `package.json` just so since it's in flux and not pushed to GitHub. Hope this helps.

<details>
<summary>package.json</summary>
<pre>
{
  "name": "@local/app",
  "description": "Sample front end (development, build)",
  "license": "MIT",
  "private": true,
  "author": "<akauppi@gmail.com>",
  "type": "module",
  "files": [
    "vitebox/dist/*"
  ],
  "exports": {
    ".": "./vitebox/dist/app.es.js"
  },
  "engines": {
    "node": ">=14.3 || ^15"
  },
  "--": "-- following stuff matters for development, only --",
  "scripts": {
    "dev": "npm run dev:local",
    "__": "",
    "test": "concurrently --kill-others --success=first -n dev,cy \"npm run dev:local\" \"wait-on http://localhost:3000 && cypress run\"",
    "  -": "",
    "cy:run": "cypress run",
    "  -- ": "",
    "dev:local": "concurrently --kill-others-on-fail -n emul,init \"firebase emulators:start --project=app --only auth,functions,firestore\" \"(export GCLOUD_PROJECT=app; npm run --silent _dev_local_init && npm run --silent _dev_local_vite)\"",
    "_dev_local_init": "wait-on http://localhost:4000 && node --experimental-json-modules --harmony-top-level-await ./local/init.js",
    "_dev_local_vite": "vite --port 3000 --mode dev_local --clearScreen false",
    "  --": "",
    "dev:online": "concurrently --kill-others-on-fail -n log,vite \"npm run --silent _dev_online_log\" \"npm run --silent _dev_online_vite\"",
    "_dev_online_log": "PORT=5050 node tools/printout.js",
    "_dev_online_vite": "wait-on http://localhost:5050 && vite --port 3001 --clearScreen false",
    "  ---": "",
    "lint": "eslint --ext .js,.vue src",
    "build": "vite build",
    "  - ": "",
    "predev:local": "npm run --silent _checkFirebaseEmulVer && (PORT=3000 npm run --silent _checkPortIsFree) && npm run --silent _genEnvLocal",
    "predev:online": "(PORT=3001 npm run --silent _checkPortIsFree) && npm run --silent _genEnvOnline",
    "//predev:online": "npm run --silent _checkFirebaseActiveProj && (PORT=3001 npm run --silent _checkPortIsFree) && npm run --silent _genEnvOnline",
    "_checkFirebaseEmulVer": "tools/firebase-ver-check.sh 8.15.0 || ( >&2 echo \"Please install 'firebase-tools' 8.15.0 or later.\n\"; exit 8)",
    "//_checkFirebaseActiveProj": "[[ $(firebase use | cat) != *\"No active project\"* ]] || ( >&2 echo \"ERROR: No active project. Please run 'firebase use <alias>' or 'firebase use --add'.\n\"; exit 1)",
    "_checkPortIsFree": "tools/port-is-free.sh $PORT || ( >&2 echo \"ERROR: Port '$PORT' is taken!\n\"; exit 6)",
    "_genEnvLocal": "node --harmony-top-level-await --experimental-json-modules tools/gen-vite-env-local.js firebase.json > vitebox/.env.dev_local",
    "_genEnvOnline": "node --harmony-top-level-await tools/gen-vite-env-online.js > vitebox/.env.development"
  },
  "dependencies": {
    "aside-keys": "^0.0.2-alpha.2",
    "firebase": "^0.900.15",
    "vue": "^3.0.6",
    "vue-router": "^4.0.4"
  },
  "devDependencies": {
    "@local/backend": "file:../backend",
    "@vitejs/plugin-vue": "^1.1.4",
    "@vue/compiler-sfc": "^3.0.6",
    "concurrently": "^6.0.0",
    "cypress": "^6.5.0",
    "eslint": "^7.20.0",
    "eslint-plugin-cypress": "^2.11.2",
    "eslint-plugin-node": "^11.1.0",
    "eslint-plugin-vue": "^7.6.0",
    "firebase-admin": "^9.5.0",
    "prettier": "^2.2.0",
    "sass": "^1.32.8",
    "vite": "^2.0.3",
    "vue-eslint-parser": "^7.5.0",
    "wait-on": "^5.2.1"
  },
  "//comments": {
    "scripts": {
      "predev:local": [
        "'firebase' CLI version (>= 8.15.0) matters for the use of auth emulation."
      ]
    },
    "devDependencies": {
      "@vue/compiler-sfc": [
        "This needs to be the same version as 'dependencies.vue' but did not find a way to enforce that."
      ],
      "prettier": [
        "tbd. why do we have it? Keep, but mention here."
      ],
      "sass": [
        "Needed as a peer dep, since we use '<style scoped lang=scss>' in the application. Otherwise, there's a RED dialog",
        "in the browser: '[plugin:vite:css] Preprocessor dependency \"sass\" not found. Did you install it?\n'"
      ]
    },
    "files": [
      "Leaving intentionally out Firebase (and tslib); index.html; those are provided by the production build."
    ],
    "engines": [
      "'>=14.3' for '—-harmony-top-level-await'",
      "'>=13.2' for ES module support",
      "'--experimental-json-modules' seems to be available for Node 12"
    ]
  }
}
</pre>
</details>

**Reply by akauppi** [reply]
This might be connected to the types exposed in the `@firebase` package:

<img width="110" alt="Näyttökuva 2021-2-26 kello 15 26 15" src="https://user-images.githubusercontent.com/1297734/109306143-6821f880-7847-11eb-81bc-03b63f68daa5.png">

Note that:

- Database has `@firebase/database/exp-types` (and shows types properly)
- Firestore doesn't have a `@firebase/firestore/exp-types` entry
- `[app|auth|database|...]-types` subpackages look like old types; should they be shipping with `@exp`!?

It would help our debugging if legacy packaging leftovers are removed from the `@exp` packaging.

**Reply by akauppi** [reply]
Firestore (0.900.15) has `node_modules/@firebase/firestore/dist/exp/index.d.ts` that has new types.

You are probably going to apply the same naming pattern to all such files, across subcomponents. Now the pattern is different from eg. `database`'s.

Furthermore:

```
$ find node_modules/@firebase/firestore -name index.d.ts | grep -v lite
node_modules/@firebase/firestore/dist/firestore/index.d.ts
node_modules/@firebase/firestore/dist/firestore/exp/index.d.ts
node_modules/@firebase/firestore/dist/exp/firestore/index.d.ts
node_modules/@firebase/firestore/dist/exp/firestore/exp/index.d.ts
node_modules/@firebase/firestore/dist/exp/packages/firestore/index.d.ts
node_modules/@firebase/firestore/dist/exp/packages/firestore/exp/index.d.ts
node_modules/@firebase/firestore/dist/exp/index.d.ts
```

Hmph 😕

**Reply by Feiyang1** [reply]
We need to clean these up. We are working on retiring all the `-types` packages and moving all types to the main packages, so you won't see the `-types` packages once the changes are merged and released.

But I'm not sure they are the issues because all that matters is which file the `typings` field points to in the `package.json`. For example, `typings` in `@firebase/firestore` points to `dist/exp/index.d.ts` which is the correct file.

I created a project using the `package.json` you provided, and typings are working correctly. What typescript version are you using?

Also, could it be an IDE specific? If it's possible, can you try a different IDE?

**Reply by akauppi** [reply]
I agree - haven’t found a glitch in the Firebase packaging that would explain this. Great to hear the packaging will get cleaned up.

Maybe we just leave this discussion open for others to pitch in?

It can definitely be IDE specific. I only use the IntelliJ ones so not eager to install others just to try. If anyone wants to do that, the instructions to set the repo up are [here](https://github.com/firebase/firebase-js-sdk/discussions/4547#discussioncomment-411243).

**Reply by akauppi** [reply]
@Feiyang1 I was trying to see where the "alpha" development happens, so I could compile and try the latest before you release it.

Did not find in this repo. Can you advice on the branch name, or is it in another, private, repository?

**Reply by Feiyang1** [reply]
It is in this repo in the master branch. Most of the new packages are located at https://github.com/firebase/firebase-js-sdk/tree/master/packages-exp. Firestore, Storage and Database are rewritten in its original location (packages/firestore, packages/storage and packages/database).

To build them, run `npm run release:exp --dry-run` at the repo root. You will need to do additional work depending on how you want to consume them. 

BTW, I'm planning to make an alpha release today, so if you can wait for a bit, you should get all updates soon.

**Reply by akauppi** [reply]
Unfortunately, this still remains in 0.900.16. 

I am willing to debug this in my end - just tell what to look for. You suggested trying with another IDE earlier, but since I'm not going to change IDE's for this, it's... nah...

I cleared the caches and restarted the IDE.  Still looks like above. 

The main problem is not the visuals,, but the inability to Cmd-B to the sources, to learn about the functions. This only works for the functions in yellow.

**Reply by akauppi** [reply]
Here's a summary of the bundle types Firebase 0.900.16 subpackages carry:

![image](https://user-images.githubusercontent.com/1297734/110777806-373fbb80-826a-11eb-9ece-1ee5bb97c5bb.png)

There isn't really much alignment. I've used "ES5" for "esm5", since there is no EcmaScript modules in ES5, only ES6. I think these are the same as "esm" in practise. 

The public-facing (no `@`) packages are not compiled for 2017. That is completely fine, since they are only proxies to the `@` implementations (I dimmed them a bit for this reason).

How is this connected to syntax highlighting...? I'm on it..

**Comment by akauppi** [comment, marked as answer]
Made a minimum reproducing repo about this: https://github.com/akauppi/firebase-syntax-highlighting-demonstrator

**Reply by akauppi** [reply]
This turned out to be my own doing..

WebStorm `Help` > `Collect Logs and Diagnostic Data` lead me to realise I had excluded all `node_modules` folders from indexing. Sure, if you don't index them, how could you provide the symbols.

I don't know why three subprojects showed symbols in the actual project. In the minimal reproduction, none did.

So. Closing. Not a general problem. Sorry for the noise!!!

**Reply by Feiyang1** [reply]
Glad you figured it out!
